### PR TITLE
[coop] Use bitfields for coop state machine state representation

### PR DESCRIPTION
**Pull request by lambdageek**
Because we have extra bits (like suspend count and the no_checkpoints bit) on
the thread state, we have to mask it to get the actual thread state enum value.  This was broken in the checked build code, previoiusly, and in general it was easy to get wrong.  Instead let's use bitfields.
But we still need access to the whole state as a 32-bit int so that we can do atomic cas

### CODE REVIEW

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Consider bitfields.

**Comment by lambdageek** [comment]
> Consider bitfields.

Yea, it's a good idea. We'd need to union things together with an int because most of the updates are done using atomics.  But for reading it will be convenient and less error-prone.

I'll rework the PR.

File: mono/utils/mono-threads.h [hunk 1]
```diff
@@ -159,6 +159,25 @@ enum {
 	ASYNC_SUSPEND_STATE_INDEX = 1,
 };
 
+typedef struct {
+#if G_BYTE_ORDER != G_BIG_ENDIAN
```
Comment (jaykrell) [inline mono/utils/mono-threads.h]: Why do you care about endian? Does anything depend on the layout?

**Comment by lambdageek** [comment]
Could I get a fresh review.  I changed the coop state representation to use bitfields.
In particular, please check that I didn't mess up the endianness.

Also, just general feedback if this is a good idea would be appreciated.  Type-punning is common, but [gcc promises to only support the most restrained uses of it](https://gcc.gnu.org/onlinedocs/gcc-4.9.3/gcc/Optimize-Options.html#Type-punning).  Maybe we were safer with masks and shifts.

File: mono/utils/mono-threads.h (hunk 1 above)
Comment (jaykrell) [reply mono/utils/mono-threads.h]: The usual pattern here is:

typedef union _Foo {
  int AsInt; // or whatever
  struct {
    int field1 : 1;
    int field2: 4;
    // etc.
 };
} Foo;

You don't need two types or to name the struct in the union.

File: mono/utils/mono-threads-state-machine.c [hunk 2]
```diff
@@ -40,7 +46,19 @@ build_thread_state (int thread_state, int suspend_count, gboolean no_safepoints)
 	g_assert (thread_state >= 0 && thread_state <= STATE_MAX);
 	no_safepoints = !!no_safepoints; // ensure it's 0 or 1
 
-	return thread_state | (suspend_count << THREAD_SUSPEND_COUNT_SHIFT) | (no_safepoints << THREAD_SUSPEND_NO_SAFEPOINTS_SHIFT);
+	MonoThreadStateMachine state;
+	state.bits.state = thread_state;
+	state.bits.no_safepoints = no_safepoints;
+	state.bits.suspend_count = suspend_count;
+	state.bits.unused = 0;
```
Comment (jaykrell) [inline mono/utils/mono-threads-state-machine.c]: I would zero the entire int first, and then fill in the used fields.

Comment (jaykrell) [reply mono/utils/mono-threads-state-machine.c]: And maybe not declared the unused bits.

**Comment by lambdageek** [comment]
Also how much should I trust compilers to pack this up for me?  do I need pragmas and attributes?

File: mono/utils/mono-threads.h (hunk 1 above)
Comment (lambdageek) [reply mono/utils/mono-threads.h]: Hm. Good point. I got rid of all the masking and shifting code, so the layout doesn't matter anymore.

**Comment by jaykrell** [comment]
No attributes, no pragma. Static_assert the size if you are nervous.

File: mono/utils/mono-threads-state-machine.c (hunk 2 above)
Comment (jaykrell) [reply mono/utils/mono-threads-state-machine.c]: Zero the overall int, then fill in the values needed.

File: mono/utils/mono-threads-state-machine.c [hunk 3]
```diff
@@ -18,19 +18,25 @@
...
+	MonoThreadStateMachine state;
+	state.raw = thread_state;
+	return state.state;
 }
 
 static int
 get_thread_suspend_count (int thread_state)
 {
-	return (thread_state & THREAD_SUSPEND_COUNT_MASK) >> THREAD_SUSPEND_COUNT_SHIFT;
+	MonoThreadStateMachine state;
```
Comment (jaykrell) [inline mono/utils/mono-threads-state-machine.c]: You can say Type var = { value };
for brevity but ok either way.

File: mono/utils/mono-threads-state-machine.c [hunk 4]
```diff
@@ -63,12 +83,25 @@ state_name (int state)
...
 }
 
-#define UNWRAP_THREAD_STATE(RAW,CUR,COUNT,BLK,INFO) do {	\
-	RAW = (INFO)->thread_state;	\
-	CUR = get_thread_state (RAW);	\
-	COUNT = get_thread_suspend_count (RAW);	\
-	BLK = get_thread_no_safepoints (RAW);	\
-} while (0)
+static void
+unwrap_thread_state (MonoThreadInfo* info,
```
Comment (jaykrell) [inline mono/utils/mono-threads-state-machine.c]: Seems like not very useful, but correct.

**Comment by imhameed** [comment]
> 
> 
> Could I get a fresh review. I changed the coop state representation to use bitfields.
> In particular, please check that I didn't mess up the endianness.

It doesn't look like any code that reads or writes to `thread_state` depends on any specific byte order or bit order, and it doesn't look like `thread_state` is ever serialized to an external byte stream (with the exception of <https://github.com/mono/mono/blob/a47ac93066b75474e9ad8aaec95c99bd16496273/mono/metadata/sgen-mono.c#L2304>, which doesn't seem important).

> Also, just general feedback if this is a good idea would be appreciated. Type-punning is common, but [gcc promises to only support the most restrained uses of it](https://gcc.gnu.org/onlinedocs/gcc-4.9.3/gcc/Optimize-Options.html#Type-punning). Maybe we were safer with masks and shifts.

I think in this case it ought to be fine, because the members of your anonymous bitfield struct have types that are all "compatible with" `int32_t`, and the strict aliasing rules have a clause that covers this (n1256 6.5 7), but I'm not sure how that interacts with bitfields.

One way to completely avoid thinking about this would be to turn `MonoThreadInfo::thread_state` into an `int32_t` (or into some new struct with a single `int32_t` member, if you want to prevent, to some degree, accidental use of `thread_state` as a plain C integer), and then write a couple of tiny, inlinable support functions to convert a `struct { int32_t state : 7; int32_t no_safepoints : 1; int32_t suspend_count : 8; int32_t zeros : 16; };` into an `int32_t` and back.

These raise more questions than they answer:
https://trust-in-soft.com/wp-content/uploads/2017/01/vmcai.pdf
http://open-std.org/jtc1/sc22/wg14/www/docs/n1409.htm
http://open-std.org/jtc1/sc22/wg14/www/docs/n1520.htm

Also: a signed integer bitfield of width 1 stores -1 and 0, and will be sign-extended into a full signed integer when accessed in any expression. That doesn't really matter for this code, though.

**Comment by jaykrell** [comment]
Gcc documentation clearly blesses this.
You don’t have to read/write same fields in union,
as long as accesses are through union,
and not taking addresses and using those.

Besides, mono likely already breaks the rules
and compiles with the flag to ask gcc 
to be reasonable.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by lewing** [comment]
@monojenkins build OS X x64 Hardened Runtime

**Pull request merged**